src/control_position.c

```diff
@@ -79,16 +86,24 @@ void control_position_update(control_position_t* cpos, uint32_t now_us) {
 		-shifted_vel_coast, shifted_vel_coast
 	);
 
+	float new_vel_output_unshifted = 0;
 	if (fabs(cpos->prop_out) > fabs(cpos->vel_output_unshifted) || cpos->prop_out * cpos->vel_output_unshifted < 0) {
-		cpos->vel_output_unshifted = linear_ramp_to(
+		new_vel_output_unshifted = linear_ramp_to(
 			cpos->vel_output_unshifted,
 			cpos->cfg->acceleration * dt,
 			cpos->prop_out
 		);
 	} else {
-		cpos->vel_output_unshifted = cpos->prop_out;
+		new_vel_output_unshifted = cpos->prop_out;
 	}
 
+	if (new_vel_output_unshifted * cpos->vel_output_unshifted < 0) {
+		cpos->changed_direction_wait_timestamp = cpos->now_us;
+		cpos->direction_changed = true;
+	}
+
+	cpos->vel_output_unshifted = new_vel_output_unshifted;
+	
 	control_position_check_target_reached(cpos);
 	cpos->vel_output = cpos->vel_output_unshifted + cpos->cfg->vel_min * signf(cpos->vel_output_unshifted);
```

Comment:
I think that vel_output should not be updated if we go into `direction_changed` mode. I would move the target reached check before line 64 (I'm not sure why it is here, but test if that breaks anything), and then update `vel_output_unshifted` in both cases, but return before updating `vel_output` if we're changing direction